### PR TITLE
ref(normalization): Split validation and normalization

### PR DESCRIPTION
**Pull request by iker-barriocanal**
This PR splits the event validation from normalization, making normalization infallible. The goal is to have two separate steps: a validation step that errors on invalid events, and an infallible normalization with the single responsibility of normalizing events. There are no functional changes in the PR.

## Design choice: a 2-step validation
Validating timestamps requires some timestamp normalization (e.g. clock drift correction), meaning the original goal of having a single validation accepting an immutable event is not easily achievable in a clean way. As a result, validation consists of two steps: a transaction validation step accepting an immutable event, and a timestamp validation step performing _minimal_ normalization. Alternatively, a small pre-validation normalization step was considered, discarded for generating too much confusion.

Further, we rely on the developer to be aware of calling the right validation function(s) and not extending the minimal normalization during validation. This has been documented in the functions.

## Future work

- Although a 2-step validation moves us closer to the goal, it's not great. This decision may be revisited as part of a larger processing pipeline changes.

#skip-changelog

### CODE REVIEW

File: relay-event-normalization/src/event.rs [hunk 1]
```diff
@@ -160,57 +129,41 @@ impl<'a> Default for NormalizationConfig<'a> {
...
     // TODO: Parts of this processor should probably be a filter so we
     // can revert some changes to ProcessingAction)
-    let mut transactions_processor = transactions::TransactionsProcessor::new(
-        config.transaction_name_config.clone(),
-        config.transaction_range.clone(),
-    );
-    transactions_processor.process_event(event, meta, ProcessingState::root())?;
+    let mut transactions_processor =
+        transactions::TransactionsProcessor::new(config.transaction_name_config.clone());
+    let _ = transactions_processor.process_event(event, meta, ProcessingState::root());
```
Comment (iker-barriocanal) [inline relay-event-normalization/src/event.rs:156]: The `TransactionsProcessor` is now infallible and doesn't reject events.

File: relay-event-normalization/src/event.rs [hunk 2]
```diff
@@ -2498,62 +2391,6 @@ mod tests {
         "###);
     }
 
-    /// Test that timestamp normalization updates a transaction's timestamps to
-    /// be acceptable, when both timestamps are similarly stale.
-    #[test]
-    fn test_accept_recent_transactions_with_stale_timestamps() {
```
Comment (iker-barriocanal) [inline relay-event-normalization/src/event.rs:2504]: These tests have been moved to `validation.rs`.

File: relay-event-normalization/src/event.rs [hunk 3]
```diff
@@ -160,57 +129,41 @@ impl<'a> Default for NormalizationConfig<'a> {
...
-        config.transaction_name_config.clone(),
-        config.transaction_range.clone(),
-    );
-    transactions_processor.process_event(event, meta, ProcessingState::root())?;
+    let mut transactions_processor =
+        transactions::TransactionsProcessor::new(config.transaction_name_config.clone());
+    let _ = transactions_processor.process_event(event, meta, ProcessingState::root());
 
     // Check for required and non-empty values
     let _ = schema::SchemaProcessor.process_event(event, meta, ProcessingState::root());
```
Comment (iker-barriocanal) [inline relay-event-normalization/src/event.rs:159]: `SchemaProcessor` only modifies the payload to add errors and remarks to `meta`, so running it after validation and timestamp normalization is fine.

File: relay-event-normalization/src/normalize/mod.rs [hunk 4]
```diff
@@ -2509,19 +2510,4 @@ mod tests {
         }
         "###);
     }
-
-    #[test]
-    fn test_reject_stale_transaction() {
```
Comment (iker-barriocanal) [inline relay-event-normalization/src/normalize/mod.rs:2514]: This test has been moved to `validation.rs`.

File: relay-event-normalization/src/transactions/processor.rs [hunk 5]
```diff
@@ -210,8 +138,10 @@ impl Processor for TransactionsProcessor<'_> {
 
         set_default_transaction_source(event);
         self.normalize_transaction_name(event);
-        self.validate_transaction(event)?;
-        end_all_spans(event)?;
+        end_all_spans(event);
+        if let Some(trace_context) = event.context_mut::<TraceContext>() {
+            trace_context.op.get_or_insert_with(|| "default".to_owned());
+        }
```
Comment (iker-barriocanal) [inline relay-event-normalization/src/transactions/processor.rs:144]: This was the only non-validation step in the old `validate_transaction`, so we keep it here.

File: relay-event-normalization/src/transactions/processor.rs [hunk 6]
```diff
@@ -438,13 +334,8 @@ fn end_all_spans(event: &mut Event) -> ProcessingResult {
                 span.timestamp.set_value(event.timestamp.value().cloned());
                 span.status = Annotated::new(SpanStatus::DeadlineExceeded);
             }
-        } else {
-            return Err(ProcessingAction::InvalidTransaction(
-                "spans must be valid in transaction event",
-            ));
```
Comment (iker-barriocanal) [inline relay-event-normalization/src/transactions/processor.rs:444]: This check is now part of the validation, and all spans at this point are valid.

File: relay-event-normalization/src/transactions/processor.rs [hunk 7]
```diff
@@ -521,25 +412,6 @@ mod tests {
         assert!(event.value().is_some());
     }
 
-    #[test]
-    fn test_discards_when_missing_timestamp() {
```
Comment (iker-barriocanal) [inline relay-event-normalization/src/transactions/processor.rs:525]: These red tests have been moved to `validation.rs`.

File: relay-event-normalization/src/transactions/processor.rs [hunk 8]
```diff
@@ -25,19 +23,12 @@ pub struct TransactionNameConfig<'r> {
...
 pub struct TransactionsProcessor<'r> {
     name_config: TransactionNameConfig<'r>,
-    timestamp_range: Option<Range<UnixTimestamp>>,
 }
 
 impl<'r> TransactionsProcessor<'r> {
     /// Creates a new `TransactionsProcessor` instance.
-    pub fn new(
-        name_config: TransactionNameConfig<'r>,
-        timestamp_range: Option<Range<UnixTimestamp>>,
```
Comment (iker-barriocanal) [inline relay-event-normalization/src/transactions/processor.rs:35]: `timestamp_range` is used during validation, so not useful in the processor anymore.

File: relay-event-normalization/src/validation.rs
Comment (iker-barriocanal) [inline relay-event-normalization/src/validation.rs:1]: Validation steps have been moved from `TransactionsProcessor`.

File: relay-server/src/services/processor.rs [hunk 9]
```diff
@@ -1095,7 +1100,11 @@ impl EnvelopeProcessorService {
             };
 
             metric!(timer(RelayTimers::EventProcessingLightNormalization), {
-                normalize_event(event, &config).map_err(|_| ProcessingError::InvalidTransaction)
+                validate_transaction(event, &tx_validation_config)?;
+                validate_event_timestamps(event, &event_validation_config)
+                    .map_err(|_| ProcessingError::InvalidTransaction)?;
+                normalize_event(event, &normalization_config);
+                Result::<(), ProcessingError>::Ok(())
```
Comment (iker-barriocanal) [inline relay-server/src/services/processor.rs:1108]: `Ok(())` isn't sufficiently explicit.

File: relay-server/src/services/processor/span/processing.rs [hunk 10]
```diff
@@ -293,9 +288,12 @@ fn normalize(
         ProcessingState::root(),
     )?;
 
+    if let Some(span) = annotated_span.value() {
+        validate_span(span)?;
+    }
     process_value(
         annotated_span,
-        &mut TransactionsProcessor::new(Default::default(), Some(timestamp_range)),
```
Comment (iker-barriocanal) [inline relay-server/src/services/processor/span/processing.rs:298]: Calling the processor with a span results in only running `process_span`, so `timestamp_range` checks were noops before. The remaining timestamp checks are done in the `validate_span` call above.
cc @jjbayer

File: relay-server/src/services/processor/span/processing.rs [hunk 11]
```diff
@@ -216,8 +217,6 @@ pub fn extract_from_event(state: &mut ProcessEnvelopeState) {
 struct NormalizeSpanConfig<'a> {
     /// The time at which the event was received in this Relay.
     pub received_at: DateTime<Utc>,
-    /// Allowed time range for transactions.
-    pub timestamp_range: std::ops::Range<UnixTimestamp>,
```
Comment (iker-barriocanal) [inline relay-server/src/services/processor/span/processing.rs:220]: With the removal of `timestamp_range` below, the property is not used anymore.

**Review by iker-barriocanal — COMMENTED** [review]
Although the diff is big there are three type of changes, for an easier review:
- Move all the validation from normalization to a separate module.
- Run validation in addition to normalization.
- Small code adaptations: removing unnecessary `unwrap`s, moving tests...

I've also added some comments. Let me know if there's anything else I can do to ease the process.

File: relay-event-normalization/src/normalize/span/tag_extraction.rs [hunk 12]
```diff
@@ -719,16 +722,21 @@ mod tests {
...
                 }
 
-                // Normalize first, to make sure that all things are correct as in the real pipeline:
-                let res = normalize_event(
+                // Validate and normalize first, to make sure that all things
+                // are correct as in the real pipeline:
+                let res = validate_transaction(&mut event, &TransactionValidationConfig::default());
+                assert!(res.is_ok());
+                let res = validate_event_timestamps(&mut event, &EventValidationConfig::default());
+                assert!(res.is_ok());
```
Comment (jjbayer) [inline relay-event-normalization/src/normalize/span/tag_extraction.rs]:
```suggestion
                // Validate and normalize first, to make sure that all things
                // are correct as in the real pipeline:
                let res = validate_transaction(&mut event, &TransactionValidationConfig::default());
                assert!(res.is_ok());
                let res = validate_event_timestamps(&mut event, &EventValidationConfig::default());
                assert!(res.is_ok());
```

This should not be necessary for the tests that use this macro. Ideally we would even call only `extract_tags` here.

File: relay-server/src/metrics_extraction/event.rs [hunk 13]
```diff
@@ -497,16 +500,21 @@ mod tests {
...
         let features = FeatureSet(BTreeSet::from([Feature::SpanMetricsExtraction]));
 
-        // Normalize first, to make sure that all things are correct as in the real pipeline:
+        // Validate and normalize first, to make sure that all things are correct as in the real pipeline:
+        let res = validate_transaction(&event, &TransactionValidationConfig::default());
+        assert!(res.is_ok());
+
+        let res = validate_event_timestamps(&mut event, &EventValidationConfig::default());
+        assert!(res.is_ok());
+
```
Comment (jjbayer) [inline relay-server/src/metrics_extraction/event.rs]: I would try to run the test without this, and remove validation if it passes. Ideally tests only run the code that they're actually trying to test. See [internal guidelines](https://www.notion.so/sentry/Relay-Code-and-Architecture-4fac481cf27741aba54d39594f1cae58?pvs=4#d8c37053239e46fbb319a06428fac18d).

```suggestion
        // Validate and normalize first, to make sure that all things are correct as in the real pipeline:
        let res = validate_transaction(&event, &TransactionValidationConfig::default());
        assert!(res.is_ok());

        let res = validate_event_timestamps(&mut event, &EventValidationConfig::default());
        assert!(res.is_ok());

```

File: relay-server/src/services/processor.rs (hunk 9 above)
Comment (jjbayer) [reply relay-server/src/services/processor.rs:1108]: Why?

File: relay-server/src/services/processor/span/processing.rs (hunk 10 above)
Comment (jjbayer) [reply relay-server/src/services/processor/span/processing.rs:298]: Thanks for this!

**Review by jjbayer — APPROVED** [review]
Please see comments before merging, but overall this looks good to me!

File: relay-server/src/services/processor.rs (hunk 9 above)
Comment (iker-barriocanal) [reply relay-server/src/services/processor.rs:1108]: The LSP doesn't recognize what type of error the `Result` is, and being explicit makes it happy.

File: relay-event-normalization/src/normalize/span/tag_extraction.rs (hunk 12 above)
Comment (iker-barriocanal) [reply relay-event-normalization/src/normalize/span/tag_extraction.rs]: Removed validation calls, but not replacing with `extract_tags` as I don't want to overload this PR.

**Pull request merged**